File: csharp/Tienda.cs
// Tienda.cs
// Tienda con inventario simple en C#
// Conceptos: interface, clase, propiedades, métodos

using System;

// INTERFACE: contrato que todo producto debe cumplir
interface IProducto {
    bool Vender(int unidades);
    void AgregarStock(int unidades);
    void Mostrar();
}

// CLASE que implementa la interface
class Producto : IProducto {
    // PROPIEDADES de solo lectura desde afuera
    public string Nombre   { get; private set; }
    public double Precio   { get; private set; }
    public int    Cantidad { get; private set; }

    // CONSTRUCTOR
    public Producto(string nombre, double precio, int cantidad) {
        Nombre   = nombre;
        Precio   = precio;
        Cantidad = cantidad;
    }

    public bool Vender(int unidades) {
        if (unidades > Cantidad) {
            Console.WriteLine($"Stock insuficiente. Solo hay {Cantidad} unidades.");
            return false;
        }
        Cantidad -= unidades;
        Console.WriteLine($"Venta de {unidades} x {Nombre} = ${Precio * unidades}");
        return true;
    }

    public void AgregarStock(int unidades) {
        Cantidad += unidades;
        Console.WriteLine($"Stock actualizado: +{unidades} {Nombre}");
    }

    public void Mostrar() {
        Console.WriteLine($"Producto : {Nombre}");
        Console.WriteLine($"Precio   : ${Precio}");
        Console.WriteLine($"Stock    : {Cantidad} unidades");
        Console.WriteLine("--------------------");
    }
}

class Program {
    static void Main(string[] args) {
        Producto p1 = new Producto("Cuaderno",  3500.0, 10);
        Producto p2 = new Producto("Lapicero",   800.0, 50);
        Producto p3 = new Producto("Mochila",  45000.0,  5);

        Console.WriteLine("=== INVENTARIO INICIAL ===");
        p1.Mostrar(); p2.Mostrar(); p3.Mostrar();

        Console.WriteLine("=== VENTAS ===");
        p1.Vender(3);
        p2.Vender(10);
        p3.Vender(10); // no hay suficiente stock

        Console.WriteLine("\n=== INVENTARIO FINAL ===");
        p1.Mostrar(); p2.Mostrar(); p3.Mostrar();
    }
}

// EJECUTAR:
// dotnet run

File: cpp/tienda.cpp

#include <iostream>
#include <string>
using namespace std;

// CLASE: representa un producto en la tienda
class Producto {
private:
    string nombre;
    double precio;
    int    cantidad;

public:
    // CONSTRUCTOR
    Producto(string nombre, double precio, int cantidad) {
        this->nombre   = nombre;
        this->precio   = precio;
        this->cantidad = cantidad;
    }

    // MÉTODO: vender una unidad
    bool vender(int unidades) {
        if (unidades > cantidad) {
            cout << "Stock insuficiente. Solo hay " << cantidad << " unidades." << endl;
            return false;
        }
        cantidad -= unidades;
        cout << "Venta de " << unidades << " x " << nombre
             << " = $" << precio * unidades << endl;
        return true;
    }

    // MÉTODO: agregar stock
    void agregar_stock(int unidades) {
        cantidad += unidades;
        cout << "Stock actualizado: +" << unidades << " " << nombre << endl;
    }

    // MÉTODO: mostrar información
    void mostrar() {
        cout << "Producto : " << nombre   << endl;
        cout << "Precio   : $" << precio  << endl;
        cout << "Stock    : " << cantidad << " unidades" << endl;
        cout << "--------------------" << endl;
    }
};

int main() {
    Producto p1("Cuaderno",   3500.0, 10);
    Producto p2("Lapicero",    800.0, 50);
    Producto p3("Mochila",   45000.0,  5);

    cout << "=== INVENTARIO INICIAL ===" << endl;
    p1.mostrar();
    p2.mostrar();
    p3.mostrar();

    cout << "=== VENTAS ===" << endl;
    p1.vender(3);
    p2.vender(10);
    p3.vender(10); // no hay suficiente stock

    cout << endl << "=== INVENTARIO FINAL ===" << endl;
    p1.mostrar();
    p2.mostrar();
    p3.mostrar();

    return 0;
}


